models/user.js

```diff
@@ -1,10 +1,148 @@
 const mongoose = require('mongoose');
+const bcrypt = require('bcrypt');
 
-const User = new mongoose.Schema({
-  name: String,
-  hash: String,
-  salt: String,
-  cookie: String,
+const {
+	Schema,
+} = mongoose;
+
+const fields = {
+	firstname: {
+		type: String,
+		required: true,
+	},
+	lastname: {
+		type: String,
+		required: true,
+	},
+	email: {
+		type: String,
+		trim: true,
+		lowercase: true,
+		required: true,
+	},
+	password: {
+		type: String,
+		required: true,
+	},
+};
+
+const user = new Schema(fields, {
+	timestamps: true,
+	toJSON: {
+		virtuals: true,
+	},
+	toObject: {
+		virtuals: true,
+	},
 });
 
-module.exports = mongoose.model('User', User);
+const blacklistFields = ['password'];
+
+user.methods.toJSON = function toJSON() {
+	const doc = this.toObject();
+	blacklistFields.forEach((field) => {
+		if (Object.hasOwnProperty.call(doc, field)) {
+			delete doc[field];
+		}
+	});
+	return doc;
+};
+
+user.pre('save', function Save(next) {
+	if (this.isNew || this.isModified('password')) {
+		this.password = bcrypt.hashSync(this.password);
+	}
+	next();
+});
+
+user.methods.verifyPassword = function verifyPassword(password) {
+	return bcrypt.compareSync(password, this.password);
+};
+
+module.exports = mongoose.model('user', user);
+const {
+      signToken,
+
+} = require('../utils/auth');
+
+
+exports.signup = (req, res, next) => {
```

Comment:
this should **not** be in models. If you want to add a way to sign up, do so, but do not add an express handler in the models directory. Modify the router in routes/api/register/users.js to do this.